homework/shared_ptr/my_shared_ptr.hpp

```diff
@@ -13,89 +13,86 @@ class my_shared_ptr
+          counter_ptr(new Counter())
+    {};
 
     my_shared_ptr(const my_shared_ptr<T>& other_shared_ptr) 
         : ptr_(other_shared_ptr.ptr_),
           counter_ptr(other_shared_ptr.counter_ptr)
     {
         counter_ptr->count_++;
-        counter_ptr->weak_count++;
     };
 
-    my_shared_ptr(my_shared_ptr<T>&& ptr_moved) :  ptr_(ptr_moved.get())
+    my_shared_ptr(my_shared_ptr<T>&& ptr_moved)
+        : ptr_(ptr_moved.ptr_), 
+        counter_ptr(ptr_moved.counter_ptr)
     {
-            counter_ptr = ptr_moved.counter_ptr;
+            ptr_moved.counter_ptr = nullptr;
             ptr_moved.ptr_ = nullptr;
     }
 
-    my_shared_ptr<T>& operator=(const my_shared_ptr<T>& other_shared_ptr) 
+   ~my_shared_ptr()
     {
-        std::cout << "operator move assigment" << "\n";
-        delete ptr_;
-        delete counter_ptr;
+        deleter();
+    }
 
-        ptr_ = other_shared_ptr.ptr_;
-        counter_ptr = other_shared_ptr.counter_ptr;
-        if (other_shared_ptr)
+    my_shared_ptr<T>& operator=(const my_shared_ptr<T>& other_shared_ptr) 
+    {
+        if(&other_shared_ptr != this)
         {
+            deleter();
+            ptr_ = other_shared_ptr.ptr_;
+            counter_ptr = other_shared_ptr.counter_ptr;
             counter_ptr->count_++;
-            counter_ptr->weak_count++;
         }
         return *this;
-    };
+    }
 
     my_shared_ptr<T>& operator=(my_shared_ptr<T>&& some_ptr)
     {
         if (&some_ptr != this)
         {
-            delete ptr_;
-            delete counter_ptr;
-
+            deleter();
             ptr_ = some_ptr.ptr_;
             counter_ptr = some_ptr.counter_ptr;
-            some_ptr = nullptr;
+            some_ptr.ptr_ = nullptr;
+            some_ptr. counter_ptr = nullptr;
         }
         return *this;
     }
 
     void reset(T* new_ptr=nullptr)
     {
-        delete ptr_;
-        counter_ptr->count_ = 0;
-        counter_ptr->weak_count = 0;
+        deleter();
         ptr_ = new_ptr;
+        counter_ptr = new Counter();
     };
 
-
-    T* get() { return ptr_; }
-    size_t use_count() { return counter_ptr->count_; }
-
-    ~my_shared_ptr()
+    void deleter() 
     {
-
-        if (counter_ptr->count_ == 0 && counter_ptr->weak_count == 0)
-        {
-            delete ptr_;
-            delete counter_ptr;
-
-        }
-        else if(counter_ptr->count_ > 0)
+        if (counter_ptr) 
         {
-            counter_ptr->count_--;
-            counter_ptr->weak_count--;
+           counter_ptr->count_--;
+           if(counter_ptr->count_ == 0)
+           {
+               delete ptr_;
+           }
+
+            if (counter_ptr->count_ == 0 && counter_ptr->weak_count == 0) 
+            {
+                delete counter_ptr;
+            }
         }
-    }
+    }  
 
-    T& operator*() { return *ptr_; };
-    T* operator->() { return ptr_; };
+    T* get() { return ptr_; }
+    size_t use_count() { return counter_ptr->count_; }
```

Comment:
```suggestion
    size_t use_count() { 
        if (ptr_) {
            return counter_ptr->count_;
        } else {
            return 0;
        }
```
This is a protection against reset operation, where you allocate a new control block with shared_refs = 1, even if you have nullptr. It may be solved in other ways too, eg. use the constructor of the control block that takes 0 as an argument.
Add checking it in ShouldResetWithNullptr test.